src/box/relay.cc

```diff
@@ -453,33 +492,50 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
 		relay_delete(relay);
 	});
 
-	/* Freeze a read view in engines. */
+	struct synchro_request req;
+	struct raft_request raft_req;
+	struct vclock limbo_vclock;
 	struct engine_join_ctx ctx;
-	engine_prepare_join_xc(&ctx);
+	ctx.array = NULL;
+	/*
+	 * Guard is defined prematurely, as it must exist in the global scope
+	 * of the function. It works only during read-view join.
+	 */
 	auto join_guard = make_scoped_guard([&] {
 		engine_complete_join(&ctx);
 	});
+	if (cursor != NULL && cursor->is_checkpoint_join) {
+		join_guard.is_active = false;
+		/** Checkpoint initial join is going to be done. */
+		int rc = relay_prepare_checkpoint_initial_join(cursor, vclock,
+							       &raft_req, &req);
```

Comment:
Why not do it in the memtx engine callback?

Reply:
Because it does much more, than callback, it recovers limbo and raft states, which are needed exclusively by relay and cannot be put inside `engine_join_ctx`, this is not a relay part. Vclock is also taken from checkpoint and not from the WAL. That's the reason, why initially everything was done as a separate function, logic differs too much

Reply:
OK, I understand now. My suggestion is to consider moving the code sending the RAFT state completely to memtx, see https://github.com/tarantool/tarantool/pull/9973/files#r1627409862.

BTW, we already have a function named `memtx_engine_recover_synchro` in `memtx_engine.cc`. The new function works only because C++ argument overloading, but we don't use that in Tarantool. Better rename it.